File: waltid-applications/waltid-wallet-demo-compose/sharedUI/src/commonMain/kotlin/id/walt/walletdemo/compose/ui/components/SharingRequestSections.kt
package id.walt.walletdemo.compose.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import id.walt.walletdemo.compose.logic.WalletDemoReaderTrust
import id.walt.walletdemo.compose.logic.WalletDemoSharingDetail
import id.walt.walletdemo.compose.logic.WalletDemoSharingEncryptionMechanism
import id.walt.walletdemo.compose.logic.WalletDemoSharingRequest
import id.walt.walletdemo.compose.logic.WalletDemoSharingRequester
import id.walt.walletdemo.compose.logic.WalletDemoSharingResponseProtection
import id.walt.walletdemo.compose.ui.WalletUiTestTags

/**
 * Renders requester identity plus the technical request facts revealed by tapping that box.
 *
 * Only the concepts [request] actually carries are rendered. A transport that has no reader
 * authentication or no requester metadata gets no such section rather than a section saying the
 * request is anonymous or unauthenticated.
 */
@Composable
internal fun SharingRequestSections(
    request: WalletDemoSharingRequest,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        VerifierMetadataCard(request)

        request.transactionData.forEach { group ->
            ClaimGroupSection(group, collapsible = false)
        }
    }
}

@Composable
private fun VerifierMetadataCard(request: WalletDemoSharingRequest) {
    val requester = request.requester
    val displayName = requester?.display?.name?.trim()?.takeIf { it.isNotEmpty() }
    val fallbackName = requester?.fallbackName?.trim()?.takeIf { it.isNotEmpty() }
    val identityName = displayName ?: fallbackName
    val verifiedOrigin = requester?.verifiedOrigin?.trim()?.takeIf { it.isNotEmpty() }
    val originIsIdentity = verifiedOrigin != null && verifiedOrigin == identityName
    val verifiedOriginDetail = verifiedOrigin
        ?.takeIf { !originIsIdentity }
        ?.let { WalletDemoSharingDetail(VERIFIED_ORIGIN_LABEL, it) }
    val requesterDetails = requester?.details.orEmpty().filter { !it.value.isNullOrBlank() }
    val technicalDetails = request.technicalDetails.filter { !it.value.isNullOrBlank() }
    val hasDetails = verifiedOriginDetail != null ||
        requesterDetails.isNotEmpty() ||
        request.readerTrust != null ||
        request.responseProtection != WalletDemoSharingResponseProtection.None ||
        technicalDetails.isNotEmpty()
    if (identityName == null && verifiedOrigin == null && !hasDetails) return

    var expanded by rememberSaveable { mutableStateOf(false) }

    ExpandableMetadataCard(
        title = "Verifier",
        expanded = expanded,
        onToggle = { expanded = !expanded },
        modifier = Modifier.testTag(WalletUiTestTags.PresentationVerifierSection),
        toggleTestTag = WalletUiTestTags.PresentationRequesterDetailsToggle,
        summary = {
            if (identityName != null) {
                MetadataIdentityRow(
                    display = requester?.display,
                    fallbackName = identityName,
                    supportingText = VERIFIED_ORIGIN_LABEL.takeIf { originIsIdentity },
                )
            } else {
                Text(
                    text = verifiedOrigin ?: "Verifier",
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = FontWeight.SemiBold,
                )
            }
        },
        details = {
            Column(
                verticalArrangement = Arrangement.spacedBy(12.dp),
                modifier = Modifier.testTag(WalletUiTestTags.PresentationRequesterDetails),
            ) {
                val identityItems = buildList {
                    verifiedOriginDetail?.let { add(MetadataDetailItem(it.label, it.value, it.linkUri)) }
                    requesterDetails.forEach { add(MetadataDetailItem(it.label, it.value, it.linkUri)) }
                }
                if (identityItems.isNotEmpty()) {
                    MetadataDetailList(identityItems)
                }
                request.readerTrust?.let {
                    if (identityItems.isNotEmpty()) MetadataRowDivider()
                    ReaderTrustDetails(it)
                }
                if (identityItems.isNotEmpty() || request.readerTrust != null) MetadataRowDivider()
                ResponseProtectionDetails(request.responseProtection)
                if (technicalDetails.isNotEmpty()) {
                    MetadataRowDivider()
                    Text(
                        "Technical request details",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.testTag(WalletUiTestTags.PresentationTechnicalDetailsSection),
                    )
                    MetadataDetailList(
                        technicalDetails.map { MetadataDetailItem(it.label, it.value, it.linkUri) },
                        modifier = Modifier.testTag(WalletUiTestTags.VerifierTechnicalDetails),
                    )
                }
            }
        },
    )
}

@Composable
private fun ReaderTrustDetails(readerTrust: WalletDemoReaderTrust) {
    val headline: String
    val explanation: String
    val identity: WalletDemoSharingDetail?
    when (readerTrust) {
        WalletDemoReaderTrust.NotAuthenticated -> {
            headline = "Reader not authenticated"
            explanation = "The request carried no reader signature, so this wallet cannot tell you who is asking."
            identity = null
        }
        WalletDemoReaderTrust.PendingVerification -> {
            headline = "Reader authentication will be verified before sharing"
            explanation = "The reader signature is checked when you share, and nothing is sent if it fails."
            identity = null
        }
        is WalletDemoReaderTrust.Untrusted -> {
            headline = "Reader identity not trusted by this wallet"
            explanation = readerTrust.reason
            identity = null
        }
        is WalletDemoReaderTrust.Trusted -> {
            headline = "Trusted reader"
            explanation = "The reader signature was verified and this wallet recognises the reader."
            identity = WalletDemoSharingDetail("Reader identity", readerTrust.readerIdentity)
        }
    }

    Column(
        verticalArrangement = Arrangement.spacedBy(4.dp),
        modifier = Modifier.testTag(WalletUiTestTags.PresentationReaderTrustSection),
    ) {
        Text(headline, style = MaterialTheme.typography.bodyMedium, fontWeight = FontWeight.Medium)
        Text(
            explanation,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        identity?.let {
            MetadataDetailList(listOf(MetadataDetailItem(it.label, it.value)))
        }
    }
}

@Composable
private fun ResponseProtectionDetails(protection: WalletDemoSharingResponseProtection) {
    Column(
        verticalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier.testTag(WalletUiTestTags.PresentationResponseProtectionSection),
    ) {
        Text(
            "Response protection",
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.SemiBold,
        )
        MetadataDetailList(
            buildList {
                add(
                    MetadataDetailItem(
                        "Message-level encryption",
                        when (protection) {
                            WalletDemoSharingResponseProtection.None -> "Not requested"
                            is WalletDemoSharingResponseProtection.Encrypted -> "Required"
                        },
                    )
                )
                if (protection is WalletDemoSharingResponseProtection.Encrypted) {
                    add(MetadataDetailItem("Encryption mechanism", protection.mechanism.displayName))
                    add(MetadataDetailItem("Key management algorithm", protection.keyManagementAlgorithm))
                    add(MetadataDetailItem("Content encryption algorithm", protection.contentEncryptionAlgorithm))
                    add(MetadataDetailItem("Verifier key ID", protection.verifierKeyId))
                    add(MetadataDetailItem("Verifier key thumbprint", protection.verifierKeyThumbprint))
                }
            }
        )
    }
}

/** Names the one requester claim the transport authenticated, wherever it is shown. */
private const val VERIFIED_ORIGIN_LABEL = "Verified website"

private val WalletDemoSharingEncryptionMechanism.displayName: String
    get() = when (this) {
        WalletDemoSharingEncryptionMechanism.Jwe -> "JWE encrypted response"
        WalletDemoSharingEncryptionMechanism.DcApiJwt -> "OpenID4VP dc_api.jwt"
        WalletDemoSharingEncryptionMechanism.AnnexCHpke -> "ISO 18013-7 Annex C HPKE"
    }

File: waltid-applications/waltid-wallet-demo-shared-ios/Sources/WalletDemoSharingUI/Components/SharingRequestSections.swift
import SwiftUI
import WalletSDK

/// Renders requester identity plus the technical request facts revealed by tapping that box.
///
/// Only the concepts ``request`` actually carries are rendered. A transport that has no reader
/// authentication or no requester metadata gets no such section rather than a section saying the
/// request is anonymous or unauthenticated.
public struct SharingRequestSections: View {
    private let request: SharingRequest

    public init(request: SharingRequest, compact: Bool = false) {
        self.request = request
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            VerifierMetadataCard(request: request)

            ForEach(request.transactionData) { group in
                ClaimGroupView(group: group, collapsible: false)
            }
        }
    }
}

private struct VerifierMetadataCard: View {
    let request: SharingRequest
    @State private var isExpanded = false

    private var requester: SharingRequester? { request.requester }

    private var verifiedOriginDetail: MetadataDetailItem? {
        guard let requester, !requester.verifiedOriginIsIdentityName,
              let origin = requester.verifiedOrigin?.presentableValue else { return nil }
        return MetadataDetailItem(label: SharingRequester.verifiedOriginLabel, value: origin)
    }

    private var requesterDetails: [MetadataDetailItem] {
        requester?.details.map { MetadataDetailItem(label: $0.label, value: $0.value, linkURI: $0.linkURI) }
            .filter(\.isVisible) ?? []
    }

    private var technicalDetails: [MetadataDetailItem] {
        request.technicalDetails
            .map { MetadataDetailItem(label: $0.label, value: $0.value, linkURI: $0.linkURI) }
            .filter(\.isVisible)
    }

    private var hasAnyContent: Bool {
        requester?.identityName != nil ||
            requester?.verifiedOrigin?.presentableValue != nil ||
            !requesterDetails.isEmpty ||
            request.readerTrust != nil ||
            request.responseProtection != .none ||
            !technicalDetails.isEmpty
    }

    var body: some View {
        if hasAnyContent {
            ExpandableMetadataCard(
                title: "Verifier",
                titleAccessibilityIdentifier: WalletAccessibilityID.presentationVerifierSection,
                toggleAccessibilityIdentifier: WalletAccessibilityID.presentationRequesterDetailsToggle,
                isExpanded: $isExpanded
            ) {
                if let identityName = requester?.identityName {
                    MetadataIdentityView(
                        display: requester?.display,
                        fallbackName: identityName,
                        supportingText: requester?.identityNameCaption
                    )
                } else {
                    Text(requester?.verifiedOrigin?.presentableValue ?? "Verifier")
                        .font(.body.weight(.semibold))
                }
            } details: {
                VStack(alignment: .leading, spacing: 12) {
                    let identityItems = [verifiedOriginDetail].compactMap { $0 } + requesterDetails
                    if !identityItems.isEmpty {
                        MetadataDetailList(items: identityItems)
                            .accessibilityIdentifier(WalletAccessibilityID.presentationRequesterDetails)
                    }
                    if let readerTrust = request.readerTrust {
                        if !identityItems.isEmpty { Divider() }
                        ReaderTrustDetails(readerTrust: readerTrust)
                    }
                    if !identityItems.isEmpty || request.readerTrust != nil { Divider() }
                    ResponseProtectionDetails(protection: request.responseProtection)
                    if !technicalDetails.isEmpty {
                        Divider()
                        Text("Technical request details")
                            .font(.caption.weight(.semibold))
                            .accessibilityIdentifier(WalletAccessibilityID.presentationTechnicalDetailsSection)
                        MetadataDetailList(items: technicalDetails)
                            .accessibilityIdentifier(WalletAccessibilityID.verifierTechnicalDetails)
                    }
                }
            }
        }
    }
}

private struct ReaderTrustDetails: View {
    let readerTrust: SharingReaderTrust

    private var headline: String {
        switch readerTrust {
        case .notAuthenticated: return "Reader not authenticated"
        case .pendingVerification: return "Reader authentication will be verified before sharing"
        case .untrusted: return "Reader identity not trusted by this wallet"
        case .trusted: return "Trusted reader"
        }
    }

    private var explanation: String {
        switch readerTrust {
        case .notAuthenticated:
            return "The request carried no reader signature, so this wallet cannot tell you who is asking."
        case .pendingVerification:
            return "The reader signature is checked when you share, and nothing is sent if it fails."
        case .untrusted(let reason):
            return reason
        case .trusted:
            return "The reader signature was verified and this wallet recognises the reader."
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(headline).font(.subheadline.weight(.medium))
            Text(explanation).font(.caption).foregroundStyle(.secondary)
            if case .trusted(let readerIdentity) = readerTrust {
                MetadataDetailList(items: [MetadataDetailItem(label: "Reader identity", value: readerIdentity)])
            }
        }
        .accessibilityIdentifier(WalletAccessibilityID.presentationReaderTrustSection)
    }
}

private struct ResponseProtectionDetails: View {
    let protection: SharingResponseProtection

    private var items: [MetadataDetailItem] {
        var items = [
            MetadataDetailItem(
                label: "Message-level encryption",
                value: protection == .none ? "Not requested" : "Required"
            ),
        ]
        if case let .encrypted(mechanism, keyManagement, contentEncryption, keyID, thumbprint) = protection {
            items += [
                MetadataDetailItem(label: "Encryption mechanism", value: mechanism.displayName),
                MetadataDetailItem(label: "Key management algorithm", value: keyManagement),
                MetadataDetailItem(label: "Content encryption algorithm", value: contentEncryption),
                MetadataDetailItem(label: "Verifier key ID", value: keyID),
                MetadataDetailItem(label: "Verifier key thumbprint", value: thumbprint),
            ]
        }
        return items
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Response protection")
                .font(.caption.weight(.semibold))
            MetadataDetailList(items: items)
        }
        .accessibilityIdentifier(WalletAccessibilityID.presentationResponseProtectionSection)
    }
}
